fix981_canonAutoFocus2

clanmills opened the pull request:
Fix for #981  This replaced PR #982.  This solution is much simpler than #982.

clanmills committed:
fix981_canonAutoFocus2

clanmills committed:
Fixed typo which caused "AFPrimaryPoint" to be missing.

derselbst reviewed (commented):
Nice job, thanks! A few issues below.

clanmills committed:
Changes based on Tom's Helpful Review @derselbst

clanmills commented on src/canonmn_int.cpp:
```diff
@@ -406,6 +406,13 @@ namespace Exiv2 {
+         {   2, N_("Single-point AF")    },
+         {   4, N_("Multi-point AF")     },
+         {   5, N_("Face Detect AF")     },
+     };
```
Done!

clanmills replied:
I changed AFMode to AFAreadMode for compatibility.

clanmills commented on src/tiffvisitor_int.cpp:
```diff
@@ -462,6 +463,53 @@ namespace Exiv2 {
+        std::string header[] =  { "AFInfoSize"     , "AFMode"          , "AFNumPoints"     , "AFValidPoints"
+                                , "AFImageWidth"   , "AFImageHeight"   , "AFWidth"         , "AFHeight"
+                                };
+        std::string points[] =  { "AFAreaWidths"   , "AFAreaHeights"   , "AFXPositions"    , "AFYPositions"  };
```
Done!

clanmills commented on src/canonmn_int.cpp:
```diff
@@ -442,11 +449,46 @@ namespace Exiv2 {
+        TagInfo(0x260b, "AFYPositions", N_("AF Y Positions"), N_("AF Y Positions"), canonId, makerTags, signedShort, -1, printValue),
+        TagInfo(0x260c, "AFPointsInFocus", N_("AF Points in Focus"), N_("AF Points in Focus"), canonId, makerTags, signedShort, -1, printValue),
+        TagInfo(0x260d, "AFPointsSelected", N_("AF Points Selected"), N_("AF Points Selected"), canonId, makerTags, signedShort, -1, printValue),
+        TagInfo(0x260e, "AFPrimaryPoint", N_("AF Primary Point"), N_("AF Primary Point"), canonId, makerTags, signedShort, -1, printValue),
```
I'm not sure that I need to do anything here.  I show the value (eg 256).  We could interpret it.  However, I think what we've got is good enough.  And for sure, I don't feel like  using my rusty Perl brains on a Friday night!

clanmills commented:
@derselbst Thanks, Tom.  And thank you for the test file.  I'll have a look at that tomorrow and add it to the test suite.  The exiv2 command-line tool can extract a `.exv` file from most images.  That's a metadata only file (it's a JPEG with no graphics).  So, we can add test images to the test suite without putting on weight.

I'm a little puzzled about your "BitMap" comments for `AFPointsInFocus`, `AFPointsSelected` and `AFPrimaryPoint`.  We can discuss it further if you feel it deserves more attention.

BTW.  This metadata cannot be edited or modified.  It's treated as part of the image.  It's bound permanently into the image.

derselbst commented:
Attached is a git patch which fixes parsing of selected AF points.

[0001-fix-decoding-of-selected-AF-fields.patch.txt](https://github.com/Exiv2/exiv2/files/3511942/0001-fix-decoding-of-selected-AF-fields.patch.txt)

derselbst commented on src/canonmn_int.cpp:
```diff
@@ -442,11 +459,46 @@ namespace Exiv2 {
+        TagInfo(0x2602, "AFNumPoints", N_("AF NumPoints"), N_("AF NumPoints"), canonId, makerTags, signedShort, -1, printValue),
+        TagInfo(0x2603, "AFValidPoints", N_("AF ValidPoints"), N_("AF ValidPoints"), canonId, makerTags, signedShort, -1, printValue),
+        TagInfo(0x2604, "AFImageWidth", N_("AF ImageWidth"), N_("AF ImageWidth"), canonId, makerTags, signedShort, -1, printValue),
+        TagInfo(0x2605, "AFImageHeight", N_("AF ImageHeight"), N_("AF ImageHeight"), canonId, makerTags, signedShort, -1, printValue),
```
One last issue: `AFImageWidth` and `AFImageHeight` is a little confusing IMO, as it actually does not really have anything to do with AF (except for the fact that it is stored in AFInfo). Phil's documentation calls it `CanonImageWidth/Height` , however he called `AFWidth/Height` `AFImageWidth/Height`.

Calling it `AFWidth/Height` below is just fine. But I'd like to suggest renaming `AFImageWidth` to `ImageWidth` and `AFImageHeight` to `ImageHeight` .They are somewhat redundant, as they are always the same as `Exif.Image.ImageWidth/Height` and `Exif.Photo.PixelX/YDimension`. I am aware that this renaming would introduce a little inconsistency to the AF tags from Nikon, so I leave it up to you whether to omit the tags, rename them or leave them as they are.

clanmills replied:
Very pleased with how this is working out.  If you'd like an unpaid job working on Exiv2, I'd be delighted to welcome you to the team!

I agree.  We should use the same names as ExifTool.  For sure the fields containing the actual image size are irrelevant and should not be there.  Let's call it CanonImageWidth/Height.

Thanks for the patch.  Nice code and works:
```
1385 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ bin/exiv2 -pa --grep af/i ~/Downloads/AAAA5440.CR2 | tail -3
Exif.Canon.AFPointsInFocus                   SShort      6  20 21 25 45 46 47
Exif.Canon.AFPointsSelected                  SShort      9  19 20 21 24 25 26 45 46 47
Exif.Canon.AFPrimaryPoint                    SShort      0  
1386 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```
I'd like to ask you to change this a little.  There's a function in TagInfo to define how to print the value of a tag.  That's where we should be printing `20 21 25...`.
```
$ grep -Hn AFPointsInFocus src/canon*.cpp
../src/canonmn_int.cpp:474:        TagInfo(0x260c, "AFPointsInFocus", N_("AF Points in Focus"), N_("AF Points in Focus"), canonId, makerTags, signedShort, -1, printValue),
../src/canonmn_int.cpp:498:44	    AFPointsInFocus	    16
$ 
```

I think we're saying that Exif.Canon.AFPointsInFocus/AFPointsSelected are 4 unsignedShorts.  When you run exiv2 with the option *-pv* (v for value) `$exiv2 -pv --grep af/i ~/Downloads/AAAA5440.CR2 | tail -3`, you see 4 unsigned shorts.  (the raw data)

When you run exiv2 with the option *-pt* (t for translated), your printing function will be called `$exiv2 -pt --grep af/i ~/Downloads/AAAA5440.CR2 | tail -3` (the human interpreted data).

There's also something weirdo about Exif.Canon.AFPrimaryPoint.  It's claiming to be of length 0.  That can't be right.

derselbst replied:
> If you'd like an unpaid job working on Exiv2, I'd be delighted to welcome you to the team!

Thanks, but I'm afraid I won't have time for this as I already have two maintainerships here at Github. Sry.

> I think we're saying that Exif.Canon.AFPointsInFocus/AFPointsSelected are 4 unsignedShorts.

In the raw exif data it's of variable length, depending how many AFPoints the camera has.

> When you run exiv2 with the option -pv (v for value) $exiv2 -pv --grep af/i ~/Downloads/AAAA5440.CR2 | tail -3, you see 4 unsigned shorts. (the raw data)
> When you run exiv2 with the option -pt (t for translated), your printing function will be called $exiv2 -pt --grep af/i ~/Downloads/AAAA5440.CR2 | tail -3 (the human interpreted data).

No matter how I run it, I always see the interpreted data.

> There's also something weirdo about Exif.Canon.AFPrimaryPoint. It's claiming to be of length 0. That can't be right.

I don't know what "PrimaryPoint" means. In my case it's always zero. Looking at the source code of exiftool, I can only assume that it's another bitmask. And when the bitmask is zero, there is no data to interpret, i.e. nothing to feed Exiv2::Value, and this results in a zero length of interpreted data. Not to be confused with the raw value.

clanmills replied:
OK.  Leave this with me.  I'm happy with your code and I'll make the changes that I have in mind and update the PR by Monday.  

`TagInfo.printFct` prints the value.  For many tags it is `printValue()` which is the same for both **-pv** (value) and **-pt** (translated).  However, some tags such as AFAreaMode do change the presentation of the data:
```
$ exiv2 -pt --grep AreaMode ~/Downloads/AAAA5440.CR2 
Exif.Canon.AFAreaMode                        SShort      1  Zone AF
$ exiv2 -pv --grep AreaMode ~/Downloads/AAAA5440.CR2 
0x2601 Canon        AFAreaMode                  SShort      1  7
$ exiv2 -pv --grep AreaMode ~/Downloads/IMG_7600.jpg 
0x2601 Canon        AFAreaMode                  SShort      1  2
$ exiv2 -pt --grep AreaMode ~/Downloads/IMG_7600.jpg 
Exif.Canon.AFAreaMode                        SShort      1  Single-point AF
$
```
Here's the code:
```
1400 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ grep AFAreaMode src/canonmn_int.cpp 
    extern const TagDetails canonAFAreaMode[] = {
        TagInfo(0x2601, "AFAreaMode", N_("AF Area Mode"), N_("AF Area Mode"), canonId, makerTags, signedShort, -1, EXV_PRINT_TAG(canonAFAreaMode)),
$ 
```

clanmills committed:
Added printBitmask(). TiffDecoder::decodeCanonAFInfo() refactored to use CanonMakerNote::tagInfo_

clanmills committed:
Added files not submitted earlier.  Updated test suite.

derselbst commented on src/tiffvisitor_int.cpp:
```diff
@@ -462,6 +463,77 @@ namespace Exiv2 {
+        	{ 0x260b , nPoints , true  }, // AFYPositions
+        	{ 0x260c , nMasks  , false }, // AFPointsInFocus
+        	{ 0x260d , nMasks  , false }, // AFPointsSelected
+        	{ 0x260e , 1       , true  }, // AFPrimaryPoint
```
... and AFPrimaryPoint must have a length of `nMasks`.

clanmills replied:
I agree with everything you've said.  Fixes submitted.

derselbst reviewed (commented):
I'm not sure if I can already comment on this since it's not Monday yet. But IMO there is no benefit in reporting the bitmasks to the user. When looking up exif data in a C++ app all I get for `"Exif.Canon.AFPointsInFocus"`   is `bitmask: (0) (512) 26 (0) (0)`. This is not really helpful as it's hard to parse. I was hoping that exiv2 just reports the focus points so that any app can simply `strtok()` and read them out.

I understand that `exiv2 -pv` should only display raw values. But in this case, strictly speaking, you should only display `Exif.Canon.AFInfo` and nothing more, because filling those buckets like `Exif.Canon.AFNumPoints` is already an interpretation of that huge AFInfo chunk.

clanmills committed:
Changes following review.  Fixed build breaker.

clanmills committed:
Fix build breaker (again!).

clanmills committed:
Fixed Linux build warnings.

clanmills commented:
If you're happy, I'll ask Luis and/or Dan to review and merge this.  Happy?

derselbst commented on src/tags_int.cpp:
```diff
@@ -2164,6 +2164,32 @@ namespace Exiv2 {
+                }
+            }
+            // if no bits are set, print the values (array of zeros)
+            if ( !comma ) printValue(os,value,metadata);
```
```suggestion
            if ( !comma ) os << 0;
```

Suggesting this change, otherwise for some files `exiv2 -pt` reports something inconsistent like:
```
Exif.Canon.AFPointsInFocus                   Short       4  0 0 0 0
Exif.Canon.AFPointsSelected                  Short       4  25
```

Here's a test file: 
[AAAA5434.exv.zip](https://github.com/Exiv2/exiv2/files/3514624/AAAA5434.exv.zip)

clanmills replied:
I made it N_("Zero") (localized string "Zero").  That's different from 0 == bit 0 is set!

I've updated the test suite.  I don't see any reason to add your file AAAA5440.exv to the test suite.  However if you think it reveals something not covered by existing 3 test files, please let me know.

derselbst replied:
> I made it N_("Zero") (localized string "Zero").

Sry, but printing a string is even more inconsistent than printing four zeros.

>  That's different from 0 == bit 0 is set!

Ok, so why not changing it to one-based indexing, so that bit 0 yields 1 ?

> I don't see any reason to add your file AAAA5440.exv to the test suite.

AAAA5440.exv has `AFPrimaryPoint` set.

derselbst reviewed (commented):
Almost, see below.

clanmills committed:
Change output from empty bitmask.

clanmills committed:
Woops.  Didn't submit this one.

clanmills commented:
We're done on this.

derselbst commented:
> We're done on this.

Missed my comment above?

clanmills commented:
I would you like to use words such as "Please" and "Thank You" when we talk.  I've put a hard week's work into your issue without a single word of appreciation or respect from you.

1) I've added AAAA5440.exv to the test suite.
2) I don't agree that "Zero" is inconsistent.
3) We could use json/python array syntax when we print the bitmask.  However, exiftool does not.
4) We shouldn't count bits starting at 1 because that exiftool doesn't.
5) You should not be parsing the output of **-pt** (translated) for two reasons:
a) You should use the exiv2 API to recover metadata.
b) If you must read the output of the exiv2 command (presumably from a script), the option **-pv** should be used.

clanmills committed:
Adding test file in which AFPrimaryPoint is not zero.

clanmills committed:
Fix test suite.

derselbst commented:
> I would you like to use words such as "Please" and "Thank You" when we talk. I've put a hard week's work into your issue without a single word of appreciation or respect from you.

Perhaps you've misunderstood my direct, nit-picky manner. I _do_ appreciate the work that you've put in this. And I believe I showed it in my first review comment and by actively contributing to this PR with code and reviews. Esp. the latter, I my opinion, is more worth than any nice words that you seem to miss in my comments. This is not meant pejorative in any way, it is just my being, sorry for that.

---

> You should not be parsing the output of -pt (translated)

That translated output however is exactly what I get when using exiv2's API. That's why I'm so nit-picky on this. All I want to say is that I don't think it's a good idea to provide a string of letters where you would only expect a series of integers. You've seen my personal suggestions above. In this particular case however, it seems even more confusing to explain to the user: "If the least significant bit is set, exiv2 returns `0`; but if no bit is set, it returns `Zero`." I was only asking you to, please, rethink about this.

> We shouldn't count bits starting at 1 because that exiftool doesn't.

As you seem to care about compatibilty to exiftool: If no bit is set, it returns `(none)` .

clanmills commented:
I am concerned about ExifTool compatibility.  Phil has done a great job and copying his designs makes sense.  I don't think it's a good idea to emit **nothing** when no bits are set.  There is a case here for using json syntax such as [ 14, 16, 29 ] for the bitmask.

I've been out to run 5k and realised that I'm not done with this.  We need code to prevent the manufactured Exif.Canon.AF* tags being written into the metadata.  That's why we can end up with two tags in the .exv file.  When the .exv is extracted, it copies the Exif.Canon.AFInfo record into the .exv AND it writes Exif.Canon.AFNumPoints etc...  The obvious fix is to add a read-only flag to TagInfo, however I don't want to touch the API for v0.27 dots.

```
$ exiv2 -pa --grep Canon.AFP test/data/test_issue_981d.exv | sort
Exif.Canon.AFPointsInFocus                   SShort      1  Zero
Exif.Canon.AFPointsInFocus                   Short       4  20,21,25,45,46,47
Exif.Canon.AFPointsSelected                  SShort      1  4,5,9
Exif.Canon.AFPointsSelected                  Short       4  19,20,21,24,25,26,45,46,47
Exif.Canon.AFPrimaryPoint                    SShort      1  13,14,15
Exif.Canon.AFPrimaryPoint                    Short       4  Zero
$ 
```

I'll have to fix this, re-extract the .exv files and modify reference output in the the python test code.  Alison and I are having an over-night trip on Tuesday/Wednesday.  If I can't fix this tonight, it's unlikely to be done until Thursday.

I want to give you a solid working fix for this issue and in return I expect you to be less abrupt in your choice of words.  I feel you are turning this into confrontation when I'm seeking cooperation.

clanmills commented:
I really don't know what do about this **nothing** business.  Let me think about this while I'm out of town.  I might ask Phil.  He's always very helpful.  It's easy to parse `Zero`.  I haven't used Yacc for a while, however it's something like:

```
bitmask : Zero
         | number-list
         ;
number-list : number
            | number-list, number
            ;
```

However, let's stop arguing  and let me think this over.  And the other matter is that the **-pt** option is provided for human use and should not be used in scripts.

clanmills commented:
We can't agree about how to represent the empty set of bits.  You'd like to see **nothing** and I want to say Zero.  There's no compromise here, so let's discuss alternatives.

1 Remove printBitmask() and report the values of the unsignedShorts.
2 Show the set as [ number, number ... ] and the set can be empty.
3 When the set is empty,  do not create the metadatum.  Exif.Canon.AFPrimaryPoint will only be present when it is not zero.  We never need to print the empty set!

Perhaps you'll think of other possibilities while Alison and I are having fun on the Isle of Wight!

derselbst commented:
> I feel you are turning this into confrontation when I'm seeking cooperation.

This has never been my intention. Apologies that you felt that way.

>  We need code to prevent the manufactured Exif.Canon.AF* tags being written into the metadata.

Good point, I missed that as well.

>  It's easy to parse Zero.

My main concern was about confusing `Zero` and `0`.

> I am concerned about ExifTool compatibility.

So, couldn't we just agree to print `(none)` just like exiftool does?

clanmills commented:
Can we clarify what you mean by `(none)`.  Do you mean empty space (nothing at all), or the string `(none)`.

If you mean empty space (nothing at all), it doesn't feel right.  However there is a precedent.  An empty ASCII string is count=0 and nothing in the data field.  For every other type of data (signed etc), the count >= 1 in the Exif specification.  Remember the count for a mask is often 4 (4 unsigned for 64 bits).  So the output says count == 4, followed by nothing at all.  

I like the idea of writing the bitmask in JSON format.  That's a lot like my original idea `bitmask: 19 20 21`.  However no need to invent anything when JSON is easy to parse in almost all languages.  So, the empty mask would be:

```Exif.Canon.PrimaryPoint  4 SHORT [ ]```

You'd know immediately you see the [, that you're to parse in JSON.

```
Exif.Canon.AFPointsInFocus                   Short       4  [ 20,21,25,45,46,47 ]
```

I think I've fixed the "never write manufactured tags" and fixed the test suite.  I'm on a car ferry at the moment.  I'll have a proper look at this on Thursday at home and then submit code.

Thank for working on this.  We're almost there.  Smile!

derselbst commented:
> Can we clarify what you mean by (none).

I mean the string `"(none)"`. Perhaps I should have written it in quotation marks right from the beginning, sry for the confusion.

clanmills committed:
Fix #985  1) printBitmask -> (none) when empty. 2) Never write manufactured tasks 3) test harness update.

clanmills commented:
Great.  We're there.  Later today, I'll close this PR and open a new short PR for team review.

clanmills commented:
This has been replaced by #988.
